agent/jvm/src/main/kotlin/io/mockk/proxy/jvm/advice/BaseAdvice.kt

```diff
@@ -42,6 +43,16 @@ internal open class BaseAdvice(
         return handler?.call()
     }
 
+    override fun isMock(instance: Any): Boolean {
+        // in order to avoid endless checks when concurrent hashmap is mocked we need to exclude handlers map explicitly
+        val castedHandlers = handlers as? JvmMockKWeakMap
```

Comment:
Suggestion: Rather than couple `BaseAdvice` with `JvmMockKWeakMap` and adding the isMock method to `JvmMockKDispatcher`, could `fun isMock(instance:Any): Boolean` be provided by an interface? If the handler is of that interface type we can then delegate the check to the handler 'MaybeMockk` implemenation?

`return (handlers as? MaybeMockK)?.isMock(instance) ?: instance !== handlers && handlers.containsKey(instance)`

Note* Take the suggestion `MaybeMockk` with a pinch of salt, a name was needed for the interface in the moment :)

Reply:
Hi, somehow missed email notification for your comments. I will try to implement your suggestion.

Reply:
<img width="1678" alt="Screenshot 2020-08-04 at 22 30 15" src="https://user-images.githubusercontent.com/10225385/89375236-1f83d180-d6a2-11ea-8839-c7f7e1a698f2.png">

Reply:
Please check the latest commit. I have introduced MockHandlerMap (Map with isMock method), just to decouple map implementation. Maybe I have misunderstood your suggestion, this way baseAdvise delegates isMock to an inner map.
Test results (same as were prior refactoring)
<img width="1678" alt="Screenshot 2020-08-04 at 22 30 15" src="https://user-images.githubusercontent.com/10225385/89375435-75f11000-d6a2-11ea-95fb-6fd57439554b.png">

@ waterpoweredmonkey